Localbackend tests

clux opened the pull request:
Based off #4 

Trying to rewrite tests. WIP.

bencord0 committed:
Move ensure_dir_exists_fresh to core/ensure.rs

bencord0 committed:
Add LocalBackend to src/storage

Local publishing and updates now work.

Basic dependency resolution:

    $ cd A
    $ lal build --release --with-version=1 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=1 to xenial

    $ lal -e xenial query A
    1

    cd ../B
    $ lal fetch
    lal::fetch: Fetch xenial A 1
    $ lal build
    ...

Now release a new version

    cd ../A
    $ lal build --release --with-version=2 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=2 to xenial

    $ lal -e xenial query A
    2
    1

Update dependencies in B

    $ cd ../B
    $ lal update A --save
    lal::update: Fetch xenial A
    lal::storage::download: Last versions for A in xenial env is {1, 2}
    lal::update: Fetch xenial A=2
    lal::update: Upgrading A from 1 to 2

bencord0 committed:
Version bump to 3.8.0

bencord0 committed:
Add examples

clux committed:
bump loggerv and expose a line number flag

clux committed:
fix bug in lal verify not checking stashed version in empty manifest

found by new tests

clux committed:
get first 7 tests working with local backend

mostly just docker with the new environment so far, but it's the
beginning of a less IO dependent test strategy.

clux committed:
make helloworld dependent on new heylib

this will fail tests, but will allow us to test dependency chains.

clux committed:
fix tests

bencord0 committed:
localstorage: Implement get_latest_version

bencord0 committed:
localstorage: handle some unwraps

bencord0 commented on configs/demo.json:
```diff
@@ -1,14 +1,12 @@
+    "local": {}
   },
   "environments": {
+    "alp36": {
```
"alpine36" or "alpine".
Is there a necessity for this specific tag? What do we do about upgrades?

clux replied:
I'll expand it, sure.

Not entirely sure if we need to be this specific either. It was intended as an abi safety by tag, but as long as the app at the top is at least newer, then we should be good. At least that was the case for ubuntu's. Alpine doesn't use glibc I guess, so maybe we can just use libraries across different alpines in newer and older alpine containers.

bencord0 commented on .gitignore:
```diff
@@ -7,4 +7,8 @@ lal.tar
 ARTIFACT
 tests/helloworld/INPUT
 tests/helloworld/.lal
+tests/heylib/INPUT
```
Maybe move the `tests/` ignores to `tests/.gitignore`.

clux committed:
cargo lock changes after rustup update

clux committed:
just call the alpine environment alpine for now

clux committed:
make tests a lot better using the local backend to publish and export

currently fails at unimplemented export.

clux committed:
fix export and query with localbackend and always use configured cache dir

this should complete the localbackend by @bencord0
tests, which now cover:

- update (not fully)
- stash
- fetch
- publish
- export
- query

seem to indicate that this is ready for primetime.

clux committed:
test lal update more rigorously

clux committed:
last missing cache use

clux committed:
old style travis build add

clux committed:
debug travis

clux committed:
fix tests and add badges

clux commented:
travis ran into an alpine image issue https://github.com/lalbuild/dockerfiles/issues/1
fixing now

clux commented:
It's all green. Merging :-)

clux commented:
localbackend released in lal@3.8.0
